lib/RecommendationProvider.js

```
@@ -0,0 +1,156 @@
+      if (this._currentRecommendation) {
+        return this._currentRecommendation;
+      }
+      return this._getRandomRecommendation();
+    }
+    // if we don't need a recommendation, don't return one
+    return null;
+  },
+
+  /**
+    * Get a new recommendation from Pocket
+    */
+  getANewRecommendation(getRecommendation) {
+    this._currentRecommendation = null;
+    this.getARecommendation(getRecommendation);
+  },
+
+  /**
+    * Randomly pick a recommendation from the allowed set of recommendations (i.e ones that are not blocked)
+    */
+  _getRandomRecommendation() {
+    const allowedRecommendations = this._recommendedContent.filter(item => !this._blockedRecommendedContent.includes(item.url));
+    if (allowedRecommendations.length > 0) {
+      let index = Math.floor(Math.random() * (allowedRecommendations.length));
+      this._currentRecommendation = Object.assign({}, {
+        url: allowedRecommendations[index].url,
+        timestamp: allowedRecommendations[index].timestamp,
+        recommended: true});
+      return this._currentRecommendation;
+    }
+    this._currentRecommendation = null;
+    return this._currentRecommendation;
+  },
+
+  /**
+    * Fetches Pocket recommendations and saves them to a global list of recommended content
+    */
+  asyncSetRecommendedContent: Task.async(function*() {
+    try {
+      let response = yield this._asyncGetRecommendationsFromPocket();
+      if (response.ok) {
+        let responseJson = yield response.json();
+        if (responseJson.urls.length > 0) {
+          // wait for the recommendation's metadata to come back (we don't want to show ugly recommendations)
+          let richRecommendations = yield this._previewProvider.getLinkMetadata(responseJson.urls);
+          this._recommendedContent = richRecommendations;
+          this._recommendedContent.forEach(item => item.timestamp = this._convertTime(item.timestamp));
+        }
+      }
```

Not sure if this matters, but there may be a potential silent failure here if `response.ok` is falsey and no error gets thrown. The `if` would skip, the `catch` wouldn't get caught, and there is no `else` statement.
It may not matter, but hey.